rr_viz/src/rr_viz/rr_viz_node.py

```diff
@@ -1,27 +1,52 @@
 #!/usr/bin/env python
 import os
 import rospy
+import rosgraph
 import threading
 import signal
 from functools import partial
 
 from PyQt5 import QtCore, QtGui, QtWidgets, uic
 from PyQt5.QtCore import *
 from PyQt5.QtGui import *
-from PyQt5.QtWidgets import *
+from PyQt5.QtWidgets import QApplication, QWidget, QInputDialog, QLineEdit, QFileDialog, QMessageBox
 from PyQt5.QtCore import pyqtSlot
-
 import managers.file_management as file_management
-
+from helpers import rr_qt_helper
 current_dir = os.path.dirname(os.path.abspath(__file__))
 Form, Base = uic.loadUiType(os.path.join(current_dir, "rr_viz.ui"))
 
 
 class MainWidget(Base, Form):
+    ross_loss_signal = QtCore.pyqtSignal(bool)
+
     def __init__(self, parent=None):
         super(self.__class__, self).__init__(parent)
         self.setupUi(self)
         self.setWindowTitle("RossRobotics RViz")
+        self.ross_loss_signal.connect(self.ros_loss_slot)
+        self.ros_loss_triggered = False
+
+    def ros_is_up(self):
+        try:
+            rosgraph.Master('/rosout').getPid()
+            return True
+        except Exception:
+            return False
+
+    def ros_loss_slot(self, flag):
```

Comment:
check line 27. ros signal  is defined as a bool signal. so on emit it takes a bool and passes it into the function it is connected to. in this case its ros_loss_slot. In this case ross_loss_signal is passed into the state_checker that on its own threa will pass  the result of ros_is_up into the signal. 

so long story short all this is so that on a separate thread you could put ros_is_up into ros_loss_slot without messing wtih QT (cuz qt doesnt like if you mess with componenets without using signals)

Reply:
@msadowski  Can you run your tests again? 
So  I have tried a few approaches: 
- Reducing amount of calls to apply_changes (the thing that publishes  all these updates bit too extreamely)
- And I tried to fully eliminating asynchronous calls to apply-changes. Since it is a - list-wide thing.  And instead just did 1 call periodically  at 10hz. 

I found that the latter did not produce good results, former did. 

on my (32gb ram) machine:
Pre fix:

Laggy:  #Waypoitns 3+
Error message: at #Waypoints >20 ish

After fix:
Laggy - same. 
error message --> i tried up to 50, no message.